src/directLineStreaming.ts

```diff
@@ -0,0 +1,351 @@
+        if (numberOfStreams !== 1) throw new Error("Expected one stream but got " + numberOfStreams)
+        const idString = await resp.streams[0].readAsString();
+        const {Id : id} = JSON.parse(idString);
+        return subscriber.next(id);
+      } catch(e) {
+          // If there is a network issue then its handled by
+          // the disconnectionHandler. Everything else can
+          // be retried
+          console.warn(e);
+          this.streamConnection.disconnect();
+          return subscriber.error(e);
+      }
+    });
+    return resp$;
+  }
+
+  private postMessageWithAttachments(message: Message) {
+    const { attachments, ...messageWithoutAttachments } = message;
+
+    let httpContentList = [];
+    return Observable.create(subscriber => {
+      return Observable.from(attachments || [])
+        .flatMap((media: Media) =>
+          Observable.ajax({
+            method: "GET",
+            url: media.contentUrl,
+            responseType: 'arraybuffer'
+          })
+            .do(ajaxResponse => {
+              let buffer = new Buffer(ajaxResponse.response);
+              let stream = new BFSE.SubscribableStream();
+              stream.write(buffer);
+              let httpContent = new BFSE.HttpContent({ type: media.contentType, contentLength: buffer.length }, stream);
+              httpContentList.push(httpContent);
+            }))
+        .count()
+        .flatMap(_ => {
+          let url = `/v3/directline/conversations/${this.conversationId}/users/${messageWithoutAttachments.from.id}/upload`;
+          let request = BFSE.StreamingRequest.create('PUT', url);
+          let activityStream = new BFSE.SubscribableStream();
+          activityStream.write(JSON.stringify(messageWithoutAttachments), 'utf-8');
+          request.addStream(new BFSE.HttpContent({ type: "application/vnd.microsoft.activity", contentLength: activityStream.length }, activityStream));
+          httpContentList.forEach(e => request.addStream(e));
+          return this.streamConnection.send(request);
+        })
+        .do(resp => {
+          if (resp.streams && resp.streams.length !== 1) {
+            subscriber.error(new Error(`Invalid stream count ${resp.streams.length}`));
+          } else {
+            resp.streams[0].readAsJson()
+              .then(({Id: id}) => {
+                subscriber.next(id)
+              })
+          }
+        })
+        .subscribe(_ => _) // force execution
+    })
+  }
+
+  private disconnectionHandler(e: any) {
+    if (this.connectionStatus$.value === ConnectionStatus.Ended) {
+      return;
+    }
+
+    if (!this.token) {
+      this.activitySubscriber.error("Token unavailable");
+      return;
+    }
+
+    this.retryCount--;
+    if (this.retryCount > 0) {
+      this.connectionStatus$.next(ConnectionStatus.Connecting);
+      setTimeout(() => {
+        this.theStreamHandler.setSubscriber(this.activitySubscriber);
+        this.connectAsync().then(_ => _);
```

Comment:
do you really need to then connectAsync here?

Reply:
turns out no :)